universal_notifications/notifications.py

```diff
@@ -57,16 +99,65 @@ def get_notification_history_details(self):
         return receiver
 
+    def check_category(self):
+        if self.category == self.PRIORITY_CATEGORY or not self.check_subscription:
+            return
+        if not self.category:
+            raise ImproperlyConfigured('Category is required', self)
+
+        notification_type = self.get_type().lower()
+        if not hasattr(settings, "UNIVERSAL_NOTIFICATIONS_CATEGORIES"):
+            raise ImproperlyConfigured("Please define UNIVERSAL_NOTIFICATIONS_CATEGORIES in your settings.py")
+
+        categories = settings.UNIVERSAL_NOTIFICATIONS_CATEGORIES.get(notification_type, {})
+        if self.category not in categories.keys():
+            raise ImproperlyConfigured("No such category for Universal Notifications: %s: %s." % (
+                self.get_type(), self.category))
+        # check if user is allowed to get notifications from this category
+        for user in self.receivers:
+            if self.category not in self.get_user_categories_for_type(user):
+                raise ImproperlyConfigured(
+                    "User is not allowed to receive notifications from '%s:%s' category"
+                    % (self.get_type(), self.category))
+
+    def verify_receivers_subscriptions(self):
+        """
+            returns new list of only receivers that are subscribed for given notification type/category
+        """
+        if not self.check_subscription or self.category == self.PRIORITY_CATEGORY:
+            return self.receivers
+
+        receivers_ids = (x.id for x in self.receivers)
+        unsubscribed_map = {}
+        for unsubscribed_user in UnsubscribedUser.objects.filter(user__in=receivers_ids):
+            unsubscribed_map[unsubscribed_user.user_id] = unsubscribed_user
+
+        filtered_receivers = []
+        ntype = self.get_type().lower()
+        for receiver in self.receivers:
+            unsubscribed_user = unsubscribed_map.get(receiver.id, None)
+            if unsubscribed_user:
+                if unsubscribed_user.unsubscribed_from_all:
+                    continue
+                unsubscribed = unsubscribed_user.unsubscribed.get(ntype, {})
+                if "all" in unsubscribed or self.category in unsubscribed:
+                    continue
+            filtered_receivers.append(receiver)
+        self.receivers = filtered_receivers
+
     def save_notifications(self, prepared_receivers):
         for receiver in prepared_receivers:
             NotificationHistory.objects.create(
                 group=self.get_type(),
                 klass=self.__class__.__name__,
                 receiver=self.format_receiver_for_notification_history(receiver),
-                details=self.get_notification_history_details()
+                details=self.get_notification_history_details(),
+                category=self.category
             )
 
     def send(self):
+        self.check_category()
+        self.verify_receivers_subscriptions()
```

Comment:
This is weird. You are checking receivers but in next line you can override it. Should default be set of subscribers instead of required to implement? or maybe first prepare and later verify?